File: research/engine/metrics.py
from __future__ import annotations

from collections import Counter
from datetime import timedelta
from statistics import mean, median
from typing import Iterable

from .models import Side, Trade


def summarize_trades(trades: Iterable[Trade]) -> dict:
    """Compute strategy-neutral performance statistics from closed trades."""
    rows = [t for t in trades if t.exit_price is not None and t.exit_time is not None]
    rs = [float(t.r_multiple) for t in rows if t.r_multiple is not None]
    wins = [r for r in rs if r > 0]
    losses = [r for r in rs if r < 0]
    gross_profit = sum(wins)
    gross_loss = abs(sum(losses))
    equity = 0.0
    peak = 0.0
    max_dd = 0.0
    for r in rs:
        equity += r
        peak = max(peak, equity)
        max_dd = max(max_dd, peak - equity)
    durations = [(t.exit_time - t.entry_time).total_seconds() for t in rows]
    long_rs = [float(t.r_multiple) for t in rows if t.side is Side.BUY and t.r_multiple is not None]
    short_rs = [float(t.r_multiple) for t in rows if t.side is Side.SELL and t.r_multiple is not None]
    max_consecutive_losses = 0
    streak = 0
    for r in rs:
        streak = streak + 1 if r < 0 else 0
        max_consecutive_losses = max(max_consecutive_losses, streak)
    return {
        "tradeCount": len(rows),
        "winRate": len(wins) / len(rs) if rs else None,
        "lossRate": len(losses) / len(rs) if rs else None,
        "expectancyR": mean(rs) if rs else None,
        "averageR": mean(rs) if rs else None,
        "medianR": median(rs) if rs else None,
        "profitFactor": gross_profit / gross_loss if gross_loss else (float("inf") if gross_profit else None),
        "maxDrawdownR": max_dd,
        "maxConsecutiveLosses": max_consecutive_losses,
        "averageHoldingSeconds": mean(durations) if durations else None,
        "longTradeCount": len(long_rs),
        "longAverageR": mean(long_rs) if long_rs else None,
        "shortTradeCount": len(short_rs),
        "shortAverageR": mean(short_rs) if short_rs else None,
    }

File: src/research/metrics.ts
import type { TradeRecord } from "../replay/trade-ledger.js";

export interface TradeMetrics {
  readonly closedTrades: number;
  readonly wins: number;
  readonly losses: number;
  readonly winRate: number;
  readonly lossRate: number;
  readonly totalR: number;
  readonly averageR: number;
  readonly medianR: number;
  readonly expectancyR: number;
  readonly grossProfitR: number;
  readonly grossLossR: number;
  readonly profitFactor: number;
  readonly maxDrawdownR: number;
  readonly maxConsecutiveLosses: number;
}

/** Descriptive metrics only; no optimization or parameter selection. */
export function calculateTradeMetrics(trades: readonly TradeRecord[]): TradeMetrics {
  const closed = trades.filter((trade) => trade.realizedR !== undefined);
  const returns = closed.map((trade) => trade.realizedR!);
  const wins = returns.filter((value) => value > 0).length;
  const losses = returns.filter((value) => value < 0).length;
  const grossProfitR = returns.filter((value) => value > 0).reduce((sum, value) => sum + value, 0);
  const grossLossR = returns.filter((value) => value < 0).reduce((sum, value) => sum + Math.abs(value), 0);
  const totalR = returns.reduce((sum, value) => sum + value, 0);
  const ordered = [...returns].sort((a, b) => a - b);
  const medianR = ordered.length === 0 ? 0 : ordered.length % 2 === 1
    ? ordered[(ordered.length - 1) / 2]
    : (ordered[ordered.length / 2 - 1] + ordered[ordered.length / 2]) / 2;

  let equity = 0;
  let peak = 0;
  let maxDrawdownR = 0;
  let consecutiveLosses = 0;
  let maxConsecutiveLosses = 0;
  for (const value of returns) {
    equity += value;
    peak = Math.max(peak, equity);
    maxDrawdownR = Math.max(maxDrawdownR, peak - equity);
    consecutiveLosses = value < 0 ? consecutiveLosses + 1 : 0;
    maxConsecutiveLosses = Math.max(maxConsecutiveLosses, consecutiveLosses);
  }

  return {
    closedTrades: closed.length,
    wins,
    losses,
    winRate: closed.length ? wins / closed.length : 0,
    lossRate: closed.length ? losses / closed.length : 0,
    totalR,
    averageR: closed.length ? totalR / closed.length : 0,
    medianR,
    expectancyR: closed.length ? totalR / closed.length : 0,
    grossProfitR,
    grossLossR,
    profitFactor: grossLossR > 0 ? grossProfitR / grossLossR : grossProfitR > 0 ? Number.POSITIVE_INFINITY : 0,
    maxDrawdownR,
    maxConsecutiveLosses,
  };
}
